// @swagger-api/apidom-parser-adapter-json/es/syntactic-analysis/PreOrderCursorChildrenIterator.js
import _defineProperty from "@babel/runtime-corejs3/helpers/esm/defineProperty";
let _Symbol$iterator;
_Symbol$iterator = Symbol.iterator;
class PreOrderCursorChildrenIterator {
  constructor(cursor) {
    _defineProperty(this, "cursor", void 0);
    this.cursor = cursor;
  }
  createNode() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      children: []
    };
  }
  *[_Symbol$iterator]() {
    // @ts-ignore
    const method = this[this.cursor.nodeType];
    const currentNode = (method || this.createNode).call(this);
    const constructor = this.constructor;
    if (this.cursor.gotoFirstChild()) {
      currentNode.children.push(...[...new constructor(this.cursor)]);
      while (this.cursor.gotoNextSibling()) {
        currentNode.children.push(...[...new constructor(this.cursor)]);
      }
      this.cursor.gotoParent();
    }
    yield currentNode;
  }
}
export default PreOrderCursorChildrenIterator;
