// @swagger-api/apidom-ns-asyncapi-2/es/refractor/visitors/async-api-2/channel-item/index.js
import stampit from 'stampit';
import { always } from 'ramda';
import { isStringElement } from '@swagger-api/apidom-core';
import ChannelItemElement from "../../../../elements/ChannelItem.js";
import FallbackVisitor from "../../FallbackVisitor.js";
import FixedFieldsVisitor from "../../generics/FixedFieldsVisitor.js";
const ChannelItemVisitor = stampit(FixedFieldsVisitor, FallbackVisitor, {
  props: {
    specPath: always(['document', 'objects', 'ChannelItem']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new ChannelItemElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = FixedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);

      // mark this ChannelItemElement with reference metadata
      if (isStringElement(this.element.$ref)) {
        this.element.classes.push('reference-element');
        this.element.setMetaProperty('referenced-element', 'channelItem');
      }
      return result;
    }
  }
});
export default ChannelItemVisitor;
