// @swagger-api/apidom-parser-adapter-json/es/syntactic-analysis/direct/CursorIterator.js
import PreOrderCursorChildrenIterator from "../PreOrderCursorChildrenIterator.js";
class CursorIterator extends PreOrderCursorChildrenIterator {
  document() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      children: []
    };
  }
  object() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  array() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  pair() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      get keyNode() {
        return this.children.find(node => node.fieldName === 'key');
      },
      get valueNode() {
        return this.children.find(node => node.fieldName === 'value');
      },
      children: []
    };
  }
  string() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      text: this.cursor.nodeText,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  number() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      text: this.cursor.nodeText,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  null() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  true() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  false() {
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      fieldName: this.cursor.currentFieldName,
      children: []
    };
  }
  ERROR() {
    const {
      currentNode
    } = this.cursor;
    return {
      type: this.cursor.nodeType,
      startPosition: this.cursor.startPosition,
      endPosition: this.cursor.endPosition,
      // @ts-ignore
      hasError: () => currentNode.hasError(),
      children: []
    };
  }
}
export default CursorIterator;
