// @swagger-api/apidom-ast/cjs/yaml/nodes/YamlStyle.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = exports.YamlStyleGroup = exports.YamlStyle = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
let YamlStyle;
exports.YamlStyle = YamlStyle;
(function (YamlStyle) {
  YamlStyle["Plain"] = "Plain";
  YamlStyle["SingleQuoted"] = "SingleQuoted";
  YamlStyle["DoubleQuoted"] = "DoubleQuoted";
  YamlStyle["Literal"] = "Literal";
  YamlStyle["Folded"] = "Folded";
  YamlStyle["Explicit"] = "Explicit";
  YamlStyle["SinglePair"] = "SinglePair";
  YamlStyle["NextLine"] = "NextLine";
  YamlStyle["InLine"] = "InLine";
})(YamlStyle || (exports.YamlStyle = YamlStyle = {}));
let YamlStyleGroup;
exports.YamlStyleGroup = YamlStyleGroup;
(function (YamlStyleGroup) {
  YamlStyleGroup["Flow"] = "Flow";
  YamlStyleGroup["Block"] = "Block";
})(YamlStyleGroup || (exports.YamlStyleGroup = YamlStyleGroup = {}));
const YamlStyleModel = (0, _stampit.default)({
  props: {
    styleGroup: null,
    style: null
  }
});
var _default = YamlStyleModel;
exports.default = _default;
