// @swagger-api/apidom-reference/cjs/util/errors/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.UnmatchedResolverError = exports.UnmatchedResolveStrategyError = exports.UnmatchedDereferenceStrategyError = exports.ResolverError = exports.PluginError = exports.ParserError = exports.NotImplementedError = exports.MaximumResolverDepthError = exports.MaximumDereferenceDepthError = exports.InvalidSelectorError = exports.DereferenceError = void 0;
var _NotImplementedError = _interopRequireDefault(require("./NotImplementedError.cjs"));
exports.NotImplementedError = _NotImplementedError.default;
var _ResolverError = _interopRequireDefault(require("./ResolverError.cjs"));
exports.ResolverError = _ResolverError.default;
var _UnmatchedResolverError = _interopRequireDefault(require("./UnmatchedResolverError.cjs"));
exports.UnmatchedResolverError = _UnmatchedResolverError.default;
var _ParserError = _interopRequireDefault(require("./ParserError.cjs"));
exports.ParserError = _ParserError.default;
var _UnmatchedResolveStrategyError = _interopRequireDefault(require("./UnmatchedResolveStrategyError.cjs"));
exports.UnmatchedResolveStrategyError = _UnmatchedResolveStrategyError.default;
var _PluginError = _interopRequireDefault(require("./PluginError.cjs"));
exports.PluginError = _PluginError.default;
var _InvalidSelectorError = _interopRequireDefault(require("./InvalidSelectorError.cjs"));
exports.InvalidSelectorError = _InvalidSelectorError.default;
var _MaximumResolverDepthError = _interopRequireDefault(require("./MaximumResolverDepthError.cjs"));
exports.MaximumResolverDepthError = _MaximumResolverDepthError.default;
var _DereferenceError = _interopRequireDefault(require("./DereferenceError.cjs"));
exports.DereferenceError = _DereferenceError.default;
var _UnmatchedDereferenceStrategyError = _interopRequireDefault(require("./UnmatchedDereferenceStrategyError.cjs"));
exports.UnmatchedDereferenceStrategyError = _UnmatchedDereferenceStrategyError.default;
var _MaximumDereferenceDepthError = _interopRequireDefault(require("./MaximumDereferenceDepthError.cjs"));
exports.MaximumDereferenceDepthError = _MaximumDereferenceDepthError.default;
