// @swagger-api/apidom-ns-asyncapi-2/cjs/refractor/visitors/async-api-2/schema/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _apidomCore = require("@swagger-api/apidom-core");
var _Schema = _interopRequireDefault(require("../../../../elements/Schema.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
var _FixedFieldsVisitor = _interopRequireDefault(require("../../generics/FixedFieldsVisitor.cjs"));
const SchemaVisitor = (0, _stampit.default)(_FixedFieldsVisitor.default, _FallbackVisitor.default, {
  props: {
    specPath: (0, _ramda.always)(['document', 'objects', 'Schema']),
    canSupportSpecificationExtensions: true
  },
  methods: {
    ObjectElement(objectElement) {
      this.element = new _Schema.default();

      // @ts-ignore
      return _FixedFieldsVisitor.default.compose.methods.ObjectElement.call(this, objectElement);
    },
    BooleanElement(booleanElement) {
      this.element = booleanElement.clone();
      this.element.classes.push('boolean-json-schema');
      return _apidomCore.BREAK;
    }
  }
});
var _default = SchemaVisitor;
exports.default = _default;
