// @swagger-api/apidom-ns-api-design-systems/cjs/refractor/specification.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _FallbackVisitor = _interopRequireDefault(require("./visitors/FallbackVisitor.cjs"));
var _index = _interopRequireDefault(require("./visitors/api-design-systems/requirement-level/index.cjs"));
var _index2 = _interopRequireDefault(require("./visitors/api-design-systems/standard-identifier/index.cjs"));
var _index3 = _interopRequireDefault(require("./visitors/api-design-systems/requirement/index.cjs"));
var _ValuesVisitor = _interopRequireDefault(require("./visitors/api-design-systems/requirement/ValuesVisitor.cjs"));
var _FollowsVisitor = _interopRequireDefault(require("./visitors/api-design-systems/requirement/FollowsVisitor.cjs"));
var _index4 = _interopRequireDefault(require("./visitors/api-design-systems/scenario/index.cjs"));
var _DescriptionVisitor = _interopRequireDefault(require("./visitors/api-design-systems/scenario/DescriptionVisitor.cjs"));
var _ThenVisitor = _interopRequireDefault(require("./visitors/api-design-systems/scenario/ThenVisitor.cjs"));
var _index5 = _interopRequireDefault(require("./visitors/api-design-systems/standard/index.cjs"));
var _NameVisitor = _interopRequireDefault(require("./visitors/api-design-systems/standard/NameVisitor.cjs"));
var _DescriptionVisitor2 = _interopRequireDefault(require("./visitors/api-design-systems/standard/DescriptionVisitor.cjs"));
var _IriVisitor = _interopRequireDefault(require("./visitors/api-design-systems/standard/IriVisitor.cjs"));
var _index6 = _interopRequireDefault(require("./visitors/api-design-systems/principle/index.cjs"));
var _NameVisitor2 = _interopRequireDefault(require("./visitors/api-design-systems/principle/NameVisitor.cjs"));
var _DescriptionVisitor3 = _interopRequireDefault(require("./visitors/api-design-systems/principle/DescriptionVisitor.cjs"));
var _IriVisitor2 = _interopRequireDefault(require("./visitors/api-design-systems/principle/IriVisitor.cjs"));
var _index7 = _interopRequireDefault(require("./visitors/api-design-systems/info/index.cjs"));
var _TitleVisitor = _interopRequireDefault(require("./visitors/api-design-systems/info/TitleVisitor.cjs"));
var _DescriptionVisitor4 = _interopRequireDefault(require("./visitors/api-design-systems/info/DescriptionVisitor.cjs"));
var _index8 = _interopRequireDefault(require("./visitors/api-design-systems/main/index.cjs"));
var _VersionVisitor = _interopRequireDefault(require("./visitors/api-design-systems/main/VersionVisitor.cjs"));
var _PrinciplesVisitor = _interopRequireDefault(require("./visitors/api-design-systems/main/PrinciplesVisitor.cjs"));
var _StandardsVisitor = _interopRequireDefault(require("./visitors/api-design-systems/main/StandardsVisitor.cjs"));
var _ScenariosVisitor = _interopRequireDefault(require("./visitors/api-design-systems/main/ScenariosVisitor.cjs"));
/**
 * API Design Systems 2021-05-07 specification elements.
 */

/**
 * Specification object allows us to have complete control over visitors
 * when traversing the ApiDOM.
 * Specification also allows us to create amended refractors from
 * existing ones by manipulating it.
 *
 * Note: Specification object allows to use absolute internal JSON pointers.
 */
const specification = {
  visitors: {
    value: _FallbackVisitor.default,
    document: {
      objects: {
        /**
         * API Design Systems 2021-05-07 specification elements.
         */
        Main: {
          $visitor: _index8.default,
          fixedFields: {
            version: _VersionVisitor.default,
            info: {
              $ref: '#/visitors/document/objects/Info'
            },
            principles: _PrinciplesVisitor.default,
            standards: _StandardsVisitor.default,
            scenarios: _ScenariosVisitor.default
          }
        },
        Info: {
          $visitor: _index7.default,
          fixedFields: {
            title: _TitleVisitor.default,
            description: _DescriptionVisitor4.default
          }
        },
        Principle: {
          $visitor: _index6.default,
          fixedFields: {
            name: _NameVisitor2.default,
            description: _DescriptionVisitor3.default,
            iri: _IriVisitor2.default,
            level: {
              $ref: '#/visitors/document/objects/RequirementLevel'
            }
          }
        },
        Standard: {
          $visitor: _index5.default,
          fixedFields: {
            name: _NameVisitor.default,
            description: _DescriptionVisitor2.default,
            level: {
              $ref: '#/visitors/document/objects/RequirementLevel'
            },
            iri: _IriVisitor.default
          }
        },
        Scenario: {
          $visitor: _index4.default,
          fixedFields: {
            description: _DescriptionVisitor.default,
            when: {
              $ref: '#/visitors/document/objects/StandardIdentifier'
            },
            then: _ThenVisitor.default
          }
        },
        Requirement: {
          $visitor: _index3.default,
          fixedFields: {
            subject: {
              $ref: '#/visitors/document/objects/StandardIdentifier'
            },
            level: {
              $ref: '#/visitors/document/objects/RequirementLevel'
            },
            values: _ValuesVisitor.default,
            follows: _FollowsVisitor.default
          }
        },
        StandardIdentifier: _index2.default,
        RequirementLevel: _index.default
      }
    }
  }
};
var _default = specification;
exports.default = _default;
