// @swagger-api/apidom-core/es/traversal/parents.js
import stampit from 'stampit';
import { visit } from "./visitor.js";
/* eslint-disable no-param-reassign */
const Visitor = stampit({
  props: {
    parent: null,
    parentEdges: null
  },
  init() {
    this.parentEdges = new WeakMap();
  },
  methods: {
    ObjectElement(objectElement) {
      this.parentEdges.set(objectElement, this.parent);
      this.parent = objectElement;
    },
    ArrayElement(arrayElement) {
      this.parentEdges.set(arrayElement, this.parent);
      this.parent = arrayElement;
    },
    MemberElement(memberElement) {
      this.parentEdges.set(memberElement, this.parent);
      this.parent = memberElement;
    },
    enter(element) {
      this.parentEdges.set(element, this.parent);
    }
  }
});

/* eslint-enable */

// computes upwards edges from every child to its parent
// find :: Element -> WeakMap<Element, Element>
const parents = element => {
  const visitor = Visitor();
  visit(element, visitor);
  return visitor.parentEdges;
};
export default parents;
