// @swagger-api/apidom-ns-asyncapi-2/cjs/refractor/visitors/async-api-2/servers/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _PatternedFieldsVisitor = _interopRequireDefault(require("../../generics/PatternedFieldsVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
var _Servers = _interopRequireDefault(require("../../../../elements/Servers.cjs"));
var _predicates = require("../../../predicates.cjs");
var _predicates2 = require("../../../../predicates.cjs");
const ServersVisitor = (0, _stampit.default)(_PatternedFieldsVisitor.default, _FallbackVisitor.default, {
  props: {
    fieldPatternPredicate: (0, _ramda.test)(/^[A-Za-z0-9_-]+$/),
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return (0, _predicates.isReferenceLikeElement)(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Server'];
    },
    canSupportSpecificationExtensions: false
  },
  init() {
    this.element = new _Servers.default();
    this.element.classes.push('servers');
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = _PatternedFieldsVisitor.default.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(_predicates2.isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'server');
      });
      return result;
    }
  }
});
var _default = ServersVisitor;
exports.default = _default;
