// @swagger-api/apidom-ns-asyncapi-2/es/refractor/visitors/async-api-2/components/ChannelBindingsVisitor.js
import stampit from 'stampit';
import ComponentsChannelBindingsElement from "../../../../elements/nces/ComponentsChannelBindings.js";
import MapVisitor from "../../generics/MapVisitor.js";
import FallbackVisitor from "../../FallbackVisitor.js";
import { isReferenceLikeElement } from "../../../predicates.js";
import { isReferenceElement } from "../../../../predicates.js";
const ChannelBindingsVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'ChannelBindings'];
    }
  },
  init() {
    this.element = new ComponentsChannelBindingsElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'channelBindings');
      });
      return result;
    }
  }
});
export default ChannelBindingsVisitor;
