// @swagger-api/apidom-ns-asyncapi-2/es/refractor/visitors/async-api-2/components/CorrelationIdsVisitor.js
import stampit from 'stampit';
import ComponentsCorrelationIDsElement from "../../../../elements/nces/ComponentsCorrelationIDs.js";
import MapVisitor from "../../generics/MapVisitor.js";
import FallbackVisitor from "../../FallbackVisitor.js";
import { isReferenceLikeElement } from "../../../predicates.js";
import { isReferenceElement } from "../../../../predicates.js";
const CorrelationIdsVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'CorrelationID'];
    }
  },
  init() {
    this.element = new ComponentsCorrelationIDsElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'correlationID');
      });
      return result;
    }
  }
});
export default CorrelationIdsVisitor;
