// @swagger-api/apidom-ns-asyncapi-2/es/refractor/visitors/async-api-2/components/ServerBindingsVisitor.js
import stampit from 'stampit';
import ComponentsServerBindingsElement from "../../../../elements/nces/ComponentsServerBindings.js";
import MapVisitor from "../../generics/MapVisitor.js";
import FallbackVisitor from "../../FallbackVisitor.js";
import { isReferenceLikeElement } from "../../../predicates.js";
import { isReferenceElement } from "../../../../predicates.js";
const ServerBindingsVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'ServerBindings'];
    }
  },
  init() {
    this.element = new ComponentsServerBindingsElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'serverBindings');
      });
      return result;
    }
  }
});
export default ServerBindingsVisitor;
