// @swagger-api/apidom-core/cjs/sexprs.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _visitor = require("./traversal/visitor.cjs");
const SymbolicExpressionsVisitor = (0, _stampit.default)({
  props: {
    nestingLevel: 0,
    result: ''
  },
  methods: {
    enter(element) {
      const {
        element: elementName
      } = element;
      const capitalizedElementName = elementName.charAt(0).toUpperCase() + elementName.slice(1);
      const indent = '  '.repeat(this.nestingLevel);
      this.result += this.nestingLevel > 0 ? '\n' : '';
      this.result += `${indent}(${capitalizedElementName}Element`;
      this.nestingLevel += 1;
    },
    leave() {
      this.nestingLevel -= 1;
      this.result += ')';
    }
  }
});

// transforms ApiDOM into S-expressions (Symbolic Expressions)
const sexprs = element => {
  const visitor = SymbolicExpressionsVisitor();
  (0, _visitor.visit)(element, visitor);
  return visitor.result;
};
var _default = sexprs;
exports.default = _default;
