// @swagger-api/apidom-ns-api-design-systems/es/refractor/registration.js
import { createRefractor } from "./index.js";
/**
 * API Design Systems 2021-05-07 specification elements.
 */
import MainElement from "../elements/Main.js";
import InfoElement from "../elements/Info.js";
import PrincipleElement from "../elements/Principle.js";
import RequirementElement from "../elements/Requirement.js";
import RequirementLevelElement from "../elements/RequirementLevel.js";
import ScenarioElement from "../elements/Scenario.js";
import StandardElement from "../elements/Standard.js";
import StandardIdentifierElement from "../elements/StandardIdentifier.js";
/**
 * API Design Systems 2021-05-07 specification elements.
 */
MainElement.refract = createRefractor(['visitors', 'document', 'objects', 'Main', '$visitor']);
InfoElement.refract = createRefractor(['visitors', 'document', 'objects', 'Info', '$visitor']);
PrincipleElement.refract = createRefractor(['visitors', 'document', 'objects', 'Principle', '$visitor']);
RequirementElement.refract = createRefractor(['visitors', 'document', 'objects', 'Requirement', '$visitor']);
RequirementLevelElement.refract = createRefractor(['visitors', 'document', 'objects', 'RequirementLevel', '$visitor']);
ScenarioElement.refract = createRefractor(['visitors', 'document', 'objects', 'Scenario', '$visitor']);
StandardElement.refract = createRefractor(['visitors', 'document', 'objects', 'Standards', '$visitor']);
StandardIdentifierElement.refract = createRefractor(['visitors', 'document', 'objects', 'StandardIdentifier', '$visitor']);
export { MainElement, InfoElement, PrincipleElement, RequirementElement, RequirementLevelElement, ScenarioElement, StandardElement, StandardIdentifierElement };
