// @swagger-api/apidom-ns-api-design-systems/es/refractor/specification.js
import FallbackVisitor from "./visitors/FallbackVisitor.js";
/**
 * API Design Systems 2021-05-07 specification elements.
 */
import RequirementLevelVisitor from "./visitors/api-design-systems/requirement-level/index.js";
import StandardIdentifierVisitor from "./visitors/api-design-systems/standard-identifier/index.js";
import RequirementVisitor from "./visitors/api-design-systems/requirement/index.js";
import RequirementValuesVisitor from "./visitors/api-design-systems/requirement/ValuesVisitor.js";
import RequirementFollowsVisitor from "./visitors/api-design-systems/requirement/FollowsVisitor.js";
import ScenarioVisitor from "./visitors/api-design-systems/scenario/index.js";
import ScenarioDescriptionVisitor from "./visitors/api-design-systems/scenario/DescriptionVisitor.js";
import ScenarioThenVisitor from "./visitors/api-design-systems/scenario/ThenVisitor.js";
import StandardVisitor from "./visitors/api-design-systems/standard/index.js";
import StandardNameVisitor from "./visitors/api-design-systems/standard/NameVisitor.js";
import StandardDescriptionVisitor from "./visitors/api-design-systems/standard/DescriptionVisitor.js";
import StandardIriVisitor from "./visitors/api-design-systems/standard/IriVisitor.js";
import PrincipleVisitor from "./visitors/api-design-systems/principle/index.js";
import PrincipleNameVisitor from "./visitors/api-design-systems/principle/NameVisitor.js";
import PrincipleDescriptionVisitor from "./visitors/api-design-systems/principle/DescriptionVisitor.js";
import PrincipleIriVisitor from "./visitors/api-design-systems/principle/IriVisitor.js";
import InfoVisitor from "./visitors/api-design-systems/info/index.js";
import InfoTitleVisitor from "./visitors/api-design-systems/info/TitleVisitor.js";
import InfoDescriptionVisitor from "./visitors/api-design-systems/info/DescriptionVisitor.js";
import MainVisitor from "./visitors/api-design-systems/main/index.js";
import MainVersionVisitor from "./visitors/api-design-systems/main/VersionVisitor.js";
import MainPrinciplesVisitor from "./visitors/api-design-systems/main/PrinciplesVisitor.js";
import MainStandardsVisitor from "./visitors/api-design-systems/main/StandardsVisitor.js";
import MainScenariosVisitor from "./visitors/api-design-systems/main/ScenariosVisitor.js";
/**
 * Specification object allows us to have complete control over visitors
 * when traversing the ApiDOM.
 * Specification also allows us to create amended refractors from
 * existing ones by manipulating it.
 *
 * Note: Specification object allows to use absolute internal JSON pointers.
 */
const specification = {
  visitors: {
    value: FallbackVisitor,
    document: {
      objects: {
        /**
         * API Design Systems 2021-05-07 specification elements.
         */
        Main: {
          $visitor: MainVisitor,
          fixedFields: {
            version: MainVersionVisitor,
            info: {
              $ref: '#/visitors/document/objects/Info'
            },
            principles: MainPrinciplesVisitor,
            standards: MainStandardsVisitor,
            scenarios: MainScenariosVisitor
          }
        },
        Info: {
          $visitor: InfoVisitor,
          fixedFields: {
            title: InfoTitleVisitor,
            description: InfoDescriptionVisitor
          }
        },
        Principle: {
          $visitor: PrincipleVisitor,
          fixedFields: {
            name: PrincipleNameVisitor,
            description: PrincipleDescriptionVisitor,
            iri: PrincipleIriVisitor,
            level: {
              $ref: '#/visitors/document/objects/RequirementLevel'
            }
          }
        },
        Standard: {
          $visitor: StandardVisitor,
          fixedFields: {
            name: StandardNameVisitor,
            description: StandardDescriptionVisitor,
            level: {
              $ref: '#/visitors/document/objects/RequirementLevel'
            },
            iri: StandardIriVisitor
          }
        },
        Scenario: {
          $visitor: ScenarioVisitor,
          fixedFields: {
            description: ScenarioDescriptionVisitor,
            when: {
              $ref: '#/visitors/document/objects/StandardIdentifier'
            },
            then: ScenarioThenVisitor
          }
        },
        Requirement: {
          $visitor: RequirementVisitor,
          fixedFields: {
            subject: {
              $ref: '#/visitors/document/objects/StandardIdentifier'
            },
            level: {
              $ref: '#/visitors/document/objects/RequirementLevel'
            },
            values: RequirementValuesVisitor,
            follows: RequirementFollowsVisitor
          }
        },
        StandardIdentifier: StandardIdentifierVisitor,
        RequirementLevel: RequirementLevelVisitor
      }
    }
  }
};
export default specification;
