// @swagger-api/apidom-parser-adapter-json/es/adapter-browser.js
import lexicalAnalysis from "./lexical-analysis/browser.js";
import syntacticAnalysisDirect from "./syntactic-analysis/direct/index.js";
import syntacticAnalysisIndirect from "./syntactic-analysis/indirect/index.js";
import { detectionRegExp } from "./adapter.js";
export { mediaTypes, namespace } from "./adapter.js";
export { detectionRegExp };
export { lexicalAnalysis, syntacticAnalysisDirect as syntacticAnalysis, syntacticAnalysisDirect, syntacticAnalysisIndirect };
export const detect = async source => {
  if (!detectionRegExp.test(source)) {
    return false;
  }
  try {
    const cst = await lexicalAnalysis(source);
    return cst.rootNode.type !== 'ERROR';
  } catch {
    return false;
  }
};
export const parse = async (source, {
  sourceMap = false,
  syntacticAnalysis = 'direct'
} = {}) => {
  const cst = await lexicalAnalysis(source);
  let apiDOM;
  if (syntacticAnalysis === 'indirect') {
    apiDOM = syntacticAnalysisIndirect(cst, {
      sourceMap
    });
  } else {
    apiDOM = syntacticAnalysisDirect(cst, {
      sourceMap
    });
  }
  return apiDOM;
};
