// @swagger-api/apidom-ns-openapi-3-0/cjs/refractor/visitors/open-api-3-0/components/RequestBodiesVisitor.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ComponentsRequestBodies = _interopRequireDefault(require("../../../../elements/nces/ComponentsRequestBodies.cjs"));
var _MapVisitor = _interopRequireDefault(require("../../generics/MapVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
var _predicates = require("../../../predicates.cjs");
var _predicates2 = require("../../../../predicates.cjs");
const RequestBodiesVisitor = (0, _stampit.default)(_MapVisitor.default, _FallbackVisitor.default, {
  props: {
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return (0, _predicates.isReferenceLikeElement)(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'RequestBody'];
    }
  },
  init() {
    this.element = new _ComponentsRequestBodies.default();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = _MapVisitor.default.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(_predicates2.isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'requestBody');
      });
      return result;
    }
  }
});
var _default = RequestBodiesVisitor;
exports.default = _default;
