// @swagger-api/apidom-ns-openapi-3-0/cjs/refractor/visitors/open-api-3-0/encoding/HeadersVisitor.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _MapVisitor = _interopRequireDefault(require("../../generics/MapVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../../FallbackVisitor.cjs"));
var _predicates = require("../../../predicates.cjs");
var _predicates2 = require("../../../../predicates.cjs");
var _EncodingHeaders = _interopRequireDefault(require("../../../../elements/nces/EncodingHeaders.cjs"));
const HeadersVisitor = (0, _stampit.default)(_MapVisitor.default, _FallbackVisitor.default, {
  props: {
    specPath: element => (0, _predicates.isReferenceLikeElement)(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Header']
  },
  init() {
    this.element = new _EncodingHeaders.default();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = _MapVisitor.default.compose.methods.ObjectElement.call(this, objectElement);

      // decorate every ReferenceElement with metadata about their referencing type
      this.element.filter(_predicates2.isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'header');
      });

      // decorate every HeaderElement with metadata about their name
      this.element.forEach((value, key) => {
        if (!(0, _predicates2.isHeaderElement)(value)) return;
        const headerName = key.toValue();
        value.setMetaProperty('headerName', headerName);
      });
      return result;
    }
  }
});
var _default = HeadersVisitor;
exports.default = _default;
