// swagger-client/lib/resolver/apidom/reference/dereference/strategies/openapi-3-1-swagger-client/visitors/parameters.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
var _compose = _interopRequireDefault(require("../utils/compose.js"));
var _toPath = _interopRequireDefault(require("../utils/to-path.js"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const ParameterMacroVisitor = (0, _compose.default)({
  init({
    parameterMacro,
    options
  }) {
    this.parameterMacro = parameterMacro;
    this.options = options;
  },
  props: {
    parameterMacro: null,
    options: null,
    macroOperation: null,
    OperationElement: {
      enter(operationElement) {
        this.macroOperation = operationElement;
      },
      leave() {
        this.macroOperation = null;
      }
    },
    ParameterElement: {
      leave(parameterElement, key, parent, path, ancestors) {
        const pojoOperation = this.macroOperation === null ? null : (0, _apidomCore.toValue)(this.macroOperation);
        const pojoParameter = (0, _apidomCore.toValue)(parameterElement);
        try {
          const macroValue = this.parameterMacro(pojoOperation, pojoParameter);
          parameterElement.set('default', macroValue);
        } catch (error) {
          var _this$options$derefer, _this$options$derefer2, _this$options$derefer3;
          const macroError = new Error(error, {
            cause: error
          });
          macroError.fullPath = (0, _toPath.default)([...ancestors, parent]);
          (_this$options$derefer = this.options.dereference.dereferenceOpts) === null || _this$options$derefer === void 0 ? void 0 : (_this$options$derefer2 = _this$options$derefer.errors) === null || _this$options$derefer2 === void 0 ? void 0 : (_this$options$derefer3 = _this$options$derefer2.push) === null || _this$options$derefer3 === void 0 ? void 0 : _this$options$derefer3.call(_this$options$derefer2, macroError);
        }
      }
    }
  }
});
var _default = ParameterMacroVisitor;
exports.default = _default;
