// @swagger-api/apidom-ns-json-schema-draft-7/es/index.js
export { isRefElement, isLinkElement, isMemberElement, isObjectElement, isArrayElement, isBooleanElement, isNullElement, isElement, isNumberElement, isStringElement } from '@swagger-api/apidom-core';
export { default as mediaTypes, JSONSchemaDraft7MediaTypes } from "./media-types.js"; // eslint-disable-next-line no-restricted-exports
export { default } from "./namespace.js";
export { default as refractorPluginReplaceEmptyElement } from "./refractor/plugins/replace-empty-element.js";
export { default as refract, createRefractor } from "./refractor/index.js";
export { default as specificationObj } from "./refractor/specification.js";
export { isJSONReferenceElement, isJSONSchemaElement, isLinkDescriptionElement } from "./predicates.js";
export { isJSONReferenceLikeElement, SpecificationVisitor, FallbackVisitor, FixedFieldsVisitor, PatternedFieldsVisitor, MapVisitor, AlternatingVisitor, ParentSchemaAwareVisitor } from '@swagger-api/apidom-ns-json-schema-draft-6';
export { keyMap, getNodeType } from "./traversal/visitor.js";
/**
 * JSON Schema Draft 7 specification elements.
 */
export { JSONSchemaElement, LinkDescriptionElement } from "./refractor/registration.js";
export { JSONReferenceElement } from '@swagger-api/apidom-ns-json-schema-draft-6';
