// @swagger-api/apidom-ns-asyncapi-2/es/refractor/visitors/async-api-2/servers/index.js
import stampit from 'stampit';
import { test } from 'ramda';
import PatternedFieldsVisitor from "../../generics/PatternedFieldsVisitor.js";
import FallbackVisitor from "../../FallbackVisitor.js";
import ServersElement from "../../../../elements/Servers.js";
import { isReferenceLikeElement } from "../../../predicates.js";
import { isReferenceElement } from "../../../../predicates.js";
const ServersVisitor = stampit(PatternedFieldsVisitor, FallbackVisitor, {
  props: {
    fieldPatternPredicate: test(/^[A-Za-z0-9_-]+$/),
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'Server'];
    },
    canSupportSpecificationExtensions: false
  },
  init() {
    this.element = new ServersElement();
    this.element.classes.push('servers');
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = PatternedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'server');
      });
      return result;
    }
  }
});
export default ServersVisitor;
