// @swagger-api/apidom-ns-json-schema-draft-6/cjs/refractor/visitors/json-schema/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _apidomCore = require("@swagger-api/apidom-core");
var _apidomNsJsonSchemaDraft = require("@swagger-api/apidom-ns-json-schema-draft-4");
var _JSONSchema = _interopRequireDefault(require("../../../elements/JSONSchema.cjs"));
const JSONSchemaVisitor = (0, _stampit.default)(_apidomNsJsonSchemaDraft.FixedFieldsVisitor, _apidomNsJsonSchemaDraft.FallbackVisitor, {
  props: {
    specPath: (0, _ramda.always)(['document', 'objects', 'JSONSchema'])
  },
  methods: {
    ObjectElement(objectElement) {
      this.element = new _JSONSchema.default();

      // @ts-ignore
      return _apidomNsJsonSchemaDraft.FixedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);
    },
    BooleanElement(booleanElement) {
      this.element = booleanElement.clone();
      this.element.classes.push('boolean-json-schema');
      return _apidomCore.BREAK;
    }
  }
});
var _default = JSONSchemaVisitor;
exports.default = _default;
