// @swagger-api/apidom-ns-api-design-systems/cjs/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.validateOpenAPI3_1 = exports.refractPluginOpenApi3_1StandardIdentifierSelectors = exports.refractPluginOpenApi3_1StandardIdentifierAccessors = exports.mediaTypes = exports.keyMap = exports.isStringElement = exports.isStandardIdentifierElement = exports.isStandardElement = exports.isScenarioElement = exports.isRequirementLevelElement = exports.isRequirementElement = exports.isRefElement = exports.isPrincipleElement = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isMainElement = exports.isLinkElement = exports.isInfoElement = exports.isElement = exports.isBooleanElement = exports.isArrayElement = exports.getNodeType = exports.default = exports.StandardIdentifierElement = exports.StandardElement = exports.ScenarioElement = exports.RequirementLevelElement = exports.RequirementElement = exports.PrincipleElement = exports.MainElement = exports.InfoElement = exports.ApiDesignSystemsMediaTypes = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.ApiDesignSystemsMediaTypes = _mediaTypes.ApiDesignSystemsMediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _standardIdentifierSelectors = _interopRequireDefault(require("./refractor/plugins/openapi-3-1/standard-identifier-selectors.cjs"));
exports.refractPluginOpenApi3_1StandardIdentifierSelectors = _standardIdentifierSelectors.default;
var _standardIdentifierAccessors = _interopRequireDefault(require("./refractor/plugins/openapi-3-1/standard-identifier-accessors.cjs"));
exports.refractPluginOpenApi3_1StandardIdentifierAccessors = _standardIdentifierAccessors.default;
var _validator = _interopRequireDefault(require("./validator/openapi-3-1/validator.cjs"));
exports.validateOpenAPI3_1 = _validator.default;
var _predicates = require("./predicates.cjs");
exports.isInfoElement = _predicates.isInfoElement;
exports.isMainElement = _predicates.isMainElement;
exports.isPrincipleElement = _predicates.isPrincipleElement;
exports.isRequirementElement = _predicates.isRequirementElement;
exports.isRequirementLevelElement = _predicates.isRequirementLevelElement;
exports.isScenarioElement = _predicates.isScenarioElement;
exports.isStandardElement = _predicates.isStandardElement;
exports.isStandardIdentifierElement = _predicates.isStandardIdentifierElement;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.MainElement = _registration.MainElement;
exports.InfoElement = _registration.InfoElement;
exports.PrincipleElement = _registration.PrincipleElement;
exports.RequirementElement = _registration.RequirementElement;
exports.RequirementLevelElement = _registration.RequirementLevelElement;
exports.ScenarioElement = _registration.ScenarioElement;
exports.StandardElement = _registration.StandardElement;
exports.StandardIdentifierElement = _registration.StandardIdentifierElement;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
