// @swagger-api/apidom-ns-json-schema-draft-4/es/index.js
export { isRefElement, isLinkElement, isMemberElement, isObjectElement, isArrayElement, isBooleanElement, isNullElement, isElement, isNumberElement, isStringElement } from '@swagger-api/apidom-core';
export { default as mediaTypes, JSONSchemaDraft4MediaTypes } from "./media-types.js"; // eslint-disable-next-line no-restricted-exports
export { default } from "./namespace.js";
export { default as refractorPluginReplaceEmptyElement } from "./refractor/plugins/replace-empty-element.js";
export { default as refract, createRefractor } from "./refractor/index.js";
export { default as specificationObj } from "./refractor/specification.js";
export { isJSONReferenceElement, isJSONSchemaElement, isLinkDescriptionElement, isMediaElement } from "./predicates.js";
export { isJSONReferenceLikeElement } from "./refractor/predicates.js";
export { default as SpecificationVisitor } from "./refractor/visitors/SpecificationVisitor.js";
export { default as FallbackVisitor } from "./refractor/visitors/FallbackVisitor.js";
export { default as FixedFieldsVisitor } from "./refractor/visitors/generics/FixedFieldsVisitor.js";
export { default as PatternedFieldsVisitor } from "./refractor/visitors/generics/PatternedFieldsVisitor.js";
export { default as MapVisitor } from "./refractor/visitors/generics/MapVisitor.js";
export { default as AlternatingVisitor } from "./refractor/visitors/generics/AlternatingVisitor.js";
export { default as ParentSchemaAwareVisitor } from "./refractor/visitors/json-schema/ParentSchemaAwareVisitor.js";
export { keyMap, getNodeType } from "./traversal/visitor.js";
/**
 * JSON Schema Draft 4 specification elements.
 */
export { JSONSchemaElement, JSONReferenceElement, MediaElement, LinkDescriptionElement } from "./refractor/registration.js";
