// @swagger-api/apidom-ns-openapi-3-0/es/refractor/visitors/open-api-3-0/components/RequestBodiesVisitor.js
import stampit from 'stampit';
import ComponentsRequestBodiesElement from "../../../../elements/nces/ComponentsRequestBodies.js";
import MapVisitor from "../../generics/MapVisitor.js";
import FallbackVisitor from "../../FallbackVisitor.js";
import { isReferenceLikeElement } from "../../../predicates.js";
import { isReferenceElement } from "../../../../predicates.js";
const RequestBodiesVisitor = stampit(MapVisitor, FallbackVisitor, {
  props: {
    specPath: element => {
      // eslint-disable-next-line no-nested-ternary
      return isReferenceLikeElement(element) ? ['document', 'objects', 'Reference'] : ['document', 'objects', 'RequestBody'];
    }
  },
  init() {
    this.element = new ComponentsRequestBodiesElement();
  },
  methods: {
    ObjectElement(objectElement) {
      // @ts-ignore
      const result = MapVisitor.compose.methods.ObjectElement.call(this, objectElement);
      this.element.filter(isReferenceElement).forEach(referenceElement => {
        referenceElement.setMetaProperty('referenced-element', 'requestBody');
      });
      return result;
    }
  }
});
export default RequestBodiesVisitor;
