// @swagger-api/apidom-ns-openapi-3-1/cjs/refractor/visitors/open-api-3-1/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _apidomNsOpenapi = require("@swagger-api/apidom-ns-openapi-3-0");
var _OpenApi = _interopRequireDefault(require("../../../elements/OpenApi3-1.cjs"));
// eslint-disable-next-line @typescript-eslint/naming-convention
const OpenApi3_1Visitor = (0, _stampit.default)(_apidomNsOpenapi.FixedFieldsVisitor, _apidomNsOpenapi.FallbackVisitor, {
  props: {
    specPath: (0, _ramda.always)(['document', 'objects', 'OpenApi']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new _OpenApi.default();
    this.openApiSemanticElement = this.element;
  },
  methods: {
    ObjectElement(objectElement) {
      this.openApiGenericElement = objectElement;

      // @ts-ignore
      return _apidomNsOpenapi.FixedFieldsVisitor.compose.methods.ObjectElement.call(this, objectElement);
    }
  }
});
var _default = OpenApi3_1Visitor;
exports.default = _default;
