// @swagger-api/apidom-ns-openapi-3-1/cjs/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.isPathItemElement = exports.isParameterLikeElement = exports.isParameterElement = exports.isOperationElement = exports.isOpenapiElement = exports.isOpenApiExtension = exports.isOpenApi3_1LikeElement = exports.isOpenApi3_1Element = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isMediaTypeElement = exports.isLinkPrimitiveElement = exports.isLinkElementExternal = exports.isLinkElement = exports.isLicenseElement = exports.isJsonSchemaDialectElement = exports.isInfoElement = exports.isExternalDocumentationElement = exports.isExampleElement = exports.isElement = exports.isContactElement = exports.isComponentsElement = exports.isCallbackElement = exports.isBooleanJsonSchemaElement = exports.isBooleanElement = exports.isArrayElement = exports.getNodeType = exports.default = exports.createToolbox = exports.XmlElement = exports.WebhooksElement = exports.TagsElement = exports.TagElement = exports.ServersElement = exports.ServerVariablesElement = exports.ServerVariableElement = exports.ServerElement = exports.SecuritySchemeElement = exports.SecurityRequirementElement = exports.SecurityElement = exports.SchemaElement = exports.ResponsesElement = exports.ResponseLinksElement = exports.ResponseHeadersElement = exports.ResponseElement = exports.ResponseContentElement = exports.RequestBodyElement = exports.RequestBodyContentElement = exports.ReferenceElement = exports.PathsElement = exports.PathItemServersElement = exports.PathItemParametersElement = exports.PathItemElement = exports.ParameterExamplesElement = exports.ParameterElement = exports.ParameterContentElement = exports.OperationTagsElement = exports.OperationServersElement = exports.OperationSecurityElement = exports.OperationParametersElement = exports.OperationElement = exports.OperationCallbacksElement = exports.OpenapiElement = exports.OpenApi3_1Element = exports.OpenAPIMediaTypes = exports.OAuthFlowsElement = exports.OAuthFlowScopesElement = exports.OAuthFlowElement = exports.MediaTypeExamplesElement = exports.MediaTypeEncodingElement = exports.MediaTypeElement = exports.LinkParametersElement = exports.LinkElement = exports.LicenseElement = exports.JsonSchemaDialectElement = exports.InfoElement = exports.HeaderExamplesElement = exports.HeaderElement = exports.HeaderContentElement = exports.ExternalDocumentationElement = exports.ExampleElement = exports.EncodingHeadersElement = exports.EncodingElement = exports.DiscriminatorMappingElement = exports.DiscriminatorElement = exports.ContactElement = exports.ComponentsSecuritySchemesElement = exports.ComponentsSchemasElement = exports.ComponentsResponsesElement = exports.ComponentsRequestBodiesElement = exports.ComponentsPathItemsElement = exports.ComponentsParametersElement = exports.ComponentsLinksElement = exports.ComponentsHeadersElement = exports.ComponentsExamplesElement = exports.ComponentsElement = exports.ComponentsCallbacksElement = exports.CallbackElement = void 0;
exports.refractorPluginReplaceEmptyElement = exports.refractorPluginNormalizeServers = exports.refractorPluginNormalizeSecurityRequirements = exports.refractorPluginNormalizeParameters = exports.refractorPluginNormalizeParameterExamples = exports.refractorPluginNormalizeOperationIds = exports.refractorPluginNormalizeHeaderExamples = exports.mediaTypes = exports.keyMap = exports.isStringElement = exports.isServerVariableElement = exports.isServerLikeElement = exports.isServerElement = exports.isSecurityRequirementElement = exports.isSchemaElement = exports.isResponsesElement = exports.isResponseLikeElement = exports.isResponseElement = exports.isRequestBodyElement = exports.isReferenceLikeElement = exports.isReferenceElementExternal = exports.isReferenceElement = exports.isRefElement = exports.isPathsElement = exports.isPathItemElementExternal = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkPrimitiveElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.OpenAPIMediaTypes = _mediaTypes.OpenAPIMediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _replaceEmptyElement = _interopRequireDefault(require("./refractor/plugins/replace-empty-element.cjs"));
exports.refractorPluginReplaceEmptyElement = _replaceEmptyElement.default;
var _normalizeParameters = _interopRequireDefault(require("./refractor/plugins/normalize-parameters.cjs"));
exports.refractorPluginNormalizeParameters = _normalizeParameters.default;
var _normalizeSecurityRequirements = _interopRequireDefault(require("./refractor/plugins/normalize-security-requirements.cjs"));
exports.refractorPluginNormalizeSecurityRequirements = _normalizeSecurityRequirements.default;
var _normalizeServers = _interopRequireDefault(require("./refractor/plugins/normalize-servers.cjs"));
exports.refractorPluginNormalizeServers = _normalizeServers.default;
var _normalizeOperationIds = _interopRequireDefault(require("./refractor/plugins/normalize-operation-ids.cjs"));
exports.refractorPluginNormalizeOperationIds = _normalizeOperationIds.default;
var _normalizeParameterExamples = _interopRequireDefault(require("./refractor/plugins/normalize-parameter-examples.cjs"));
exports.refractorPluginNormalizeParameterExamples = _normalizeParameterExamples.default;
var _normalizeHeaderExamples = _interopRequireDefault(require("./refractor/plugins/normalize-header-examples.cjs"));
exports.refractorPluginNormalizeHeaderExamples = _normalizeHeaderExamples.default;
var _toolbox = _interopRequireDefault(require("./refractor/toolbox.cjs"));
exports.createToolbox = _toolbox.default;
var _predicates = require("./predicates.cjs");
exports.isCallbackElement = _predicates.isCallbackElement;
exports.isComponentsElement = _predicates.isComponentsElement;
exports.isInfoElement = _predicates.isInfoElement;
exports.isJsonSchemaDialectElement = _predicates.isJsonSchemaDialectElement;
exports.isLicenseElement = _predicates.isLicenseElement;
exports.isOpenapiElement = _predicates.isOpenapiElement;
exports.isOpenApi3_1Element = _predicates.isOpenApi3_1Element;
exports.isOperationElement = _predicates.isOperationElement;
exports.isParameterElement = _predicates.isParameterElement;
exports.isPathItemElement = _predicates.isPathItemElement;
exports.isPathItemElementExternal = _predicates.isPathItemElementExternal;
exports.isReferenceElement = _predicates.isReferenceElement;
exports.isReferenceElementExternal = _predicates.isReferenceElementExternal;
exports.isResponseElement = _predicates.isResponseElement;
exports.isResponsesElement = _predicates.isResponsesElement;
exports.isSchemaElement = _predicates.isSchemaElement;
exports.isBooleanJsonSchemaElement = _predicates.isBooleanJsonSchemaElement;
exports.isMediaTypeElement = _predicates.isMediaTypeElement;
exports.isServerElement = _predicates.isServerElement;
exports.isSecurityRequirementElement = _predicates.isSecurityRequirementElement;
exports.isExternalDocumentationElement = _predicates.isExternalDocumentationElement;
exports.isServerVariableElement = _predicates.isServerVariableElement;
exports.isContactElement = _predicates.isContactElement;
exports.isExampleElement = _predicates.isExampleElement;
exports.isLinkElement = _predicates.isLinkElement;
exports.isLinkElementExternal = _predicates.isLinkElementExternal;
exports.isRequestBodyElement = _predicates.isRequestBodyElement;
exports.isPathsElement = _predicates.isPathsElement;
var _predicates2 = require("./refractor/predicates.cjs");
exports.isOpenApi3_1LikeElement = _predicates2.isOpenApi3_1LikeElement;
var _apidomNsOpenapi = require("@swagger-api/apidom-ns-openapi-3-0");
exports.isParameterLikeElement = _apidomNsOpenapi.isParameterLikeElement;
exports.isReferenceLikeElement = _apidomNsOpenapi.isReferenceLikeElement;
exports.isResponseLikeElement = _apidomNsOpenapi.isResponseLikeElement;
exports.isOpenApiExtension = _apidomNsOpenapi.isOpenApiExtension;
exports.isServerLikeElement = _apidomNsOpenapi.isServerLikeElement;
exports.ComponentsCallbacksElement = _apidomNsOpenapi.ComponentsCallbacksElement;
exports.ComponentsExamplesElement = _apidomNsOpenapi.ComponentsExamplesElement;
exports.ComponentsHeadersElement = _apidomNsOpenapi.ComponentsHeadersElement;
exports.ComponentsLinksElement = _apidomNsOpenapi.ComponentsLinksElement;
exports.ComponentsParametersElement = _apidomNsOpenapi.ComponentsParametersElement;
exports.ComponentsRequestBodiesElement = _apidomNsOpenapi.ComponentsRequestBodiesElement;
exports.ComponentsResponsesElement = _apidomNsOpenapi.ComponentsResponsesElement;
exports.ComponentsSchemasElement = _apidomNsOpenapi.ComponentsSchemasElement;
exports.ComponentsSecuritySchemesElement = _apidomNsOpenapi.ComponentsSecuritySchemesElement;
exports.DiscriminatorMappingElement = _apidomNsOpenapi.DiscriminatorMappingElement;
exports.EncodingHeadersElement = _apidomNsOpenapi.EncodingHeadersElement;
exports.HeaderContentElement = _apidomNsOpenapi.HeaderContentElement;
exports.HeaderExamplesElement = _apidomNsOpenapi.HeaderExamplesElement;
exports.LinkParametersElement = _apidomNsOpenapi.LinkParametersElement;
exports.MediaTypeEncodingElement = _apidomNsOpenapi.MediaTypeEncodingElement;
exports.MediaTypeExamplesElement = _apidomNsOpenapi.MediaTypeExamplesElement;
exports.OAuthFlowScopesElement = _apidomNsOpenapi.OAuthFlowScopesElement;
exports.OperationCallbacksElement = _apidomNsOpenapi.OperationCallbacksElement;
exports.OperationParametersElement = _apidomNsOpenapi.OperationParametersElement;
exports.OperationSecurityElement = _apidomNsOpenapi.OperationSecurityElement;
exports.OperationServersElement = _apidomNsOpenapi.OperationServersElement;
exports.OperationTagsElement = _apidomNsOpenapi.OperationTagsElement;
exports.ParameterContentElement = _apidomNsOpenapi.ParameterContentElement;
exports.ParameterExamplesElement = _apidomNsOpenapi.ParameterExamplesElement;
exports.PathItemParametersElement = _apidomNsOpenapi.PathItemParametersElement;
exports.PathItemServersElement = _apidomNsOpenapi.PathItemServersElement;
exports.RequestBodyContentElement = _apidomNsOpenapi.RequestBodyContentElement;
exports.ResponseContentElement = _apidomNsOpenapi.ResponseContentElement;
exports.ResponseHeadersElement = _apidomNsOpenapi.ResponseHeadersElement;
exports.ResponseLinksElement = _apidomNsOpenapi.ResponseLinksElement;
exports.SecurityElement = _apidomNsOpenapi.SecurityElement;
exports.ServersElement = _apidomNsOpenapi.ServersElement;
exports.ServerVariablesElement = _apidomNsOpenapi.ServerVariablesElement;
exports.TagsElement = _apidomNsOpenapi.TagsElement;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.CallbackElement = _registration.CallbackElement;
exports.ComponentsElement = _registration.ComponentsElement;
exports.ContactElement = _registration.ContactElement;
exports.DiscriminatorElement = _registration.DiscriminatorElement;
exports.EncodingElement = _registration.EncodingElement;
exports.ExampleElement = _registration.ExampleElement;
exports.ExternalDocumentationElement = _registration.ExternalDocumentationElement;
exports.HeaderElement = _registration.HeaderElement;
exports.InfoElement = _registration.InfoElement;
exports.JsonSchemaDialectElement = _registration.JsonSchemaDialectElement;
exports.LicenseElement = _registration.LicenseElement;
exports.LinkElement = _registration.LinkElement;
exports.MediaTypeElement = _registration.MediaTypeElement;
exports.OAuthFlowElement = _registration.OAuthFlowElement;
exports.OAuthFlowsElement = _registration.OAuthFlowsElement;
exports.OpenapiElement = _registration.OpenapiElement;
exports.OpenApi3_1Element = _registration.OpenApi3_1Element;
exports.OperationElement = _registration.OperationElement;
exports.ParameterElement = _registration.ParameterElement;
exports.PathItemElement = _registration.PathItemElement;
exports.PathsElement = _registration.PathsElement;
exports.ReferenceElement = _registration.ReferenceElement;
exports.RequestBodyElement = _registration.RequestBodyElement;
exports.ResponseElement = _registration.ResponseElement;
exports.ResponsesElement = _registration.ResponsesElement;
exports.SchemaElement = _registration.SchemaElement;
exports.SecurityRequirementElement = _registration.SecurityRequirementElement;
exports.SecuritySchemeElement = _registration.SecuritySchemeElement;
exports.ServerElement = _registration.ServerElement;
exports.ServerVariableElement = _registration.ServerVariableElement;
exports.TagElement = _registration.TagElement;
exports.XmlElement = _registration.XmlElement;
var _ComponentsPathItems = _interopRequireDefault(require("./elements/nces/ComponentsPathItems.cjs"));
exports.ComponentsPathItemsElement = _ComponentsPathItems.default;
var _Webhooks = _interopRequireDefault(require("./elements/nces/Webhooks.cjs"));
exports.WebhooksElement = _Webhooks.default;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
