// @swagger-api/apidom-ns-asyncapi-2/es/refractor/predicates.js
import { isStringElement, isObjectElement } from '@swagger-api/apidom-core';
import { startsWith } from 'ramda';
export const isAsyncApi2LikeElement = element => {
  return isObjectElement(element) &&
  // @ts-ignore
  element.hasKey('asyncapi') &&
  // @ts-ignore
  element.hasKey('info') &&
  // @ts-ignore
  element.hasKey('channels');
};
export const isParameterLikeElement = element => {
  // @ts-ignore
  return isObjectElement(element);
};
export const isReferenceLikeElement = element => {
  // @ts-ignore
  return isObjectElement(element) && element.hasKey('$ref');
};
export const isSchemaLikeElement = element => {
  // @ts-ignore
  return isObjectElement(element);
};
export const isSecurityRequirementLikeElement = element => {
  return isObjectElement(element);
};
export const isServerLikeElement = element => {
  // @ts-ignore
  return isObjectElement(element) && element.hasKey('url') && element.hasKey('protocol');
};
export const isAsyncApiExtension = element => {
  // @ts-ignore
  return isStringElement(element.key) && startsWith('x-', element.key.toValue());
};
