// @swagger-api/apidom-core/cjs/traversal/findAtOffset.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _ramdaAdjunct = require("ramda-adjunct");
var _index = require("../predicates/index.cjs");
var _visitor = require("./visitor.cjs");
const Visitor = (0, _stampit.default)({
  props: {
    result: [],
    offset: 0,
    includeRightBound: false
  },
  // @ts-ignore
  init({
    offset = this.offset,
    includeRightBound = this.includeRightBound
  }) {
    this.result = [];
    this.offset = offset;
    this.includeRightBound = includeRightBound;
  },
  methods: {
    enter(element) {
      if (!(0, _index.hasElementSourceMap)(element)) {
        return undefined; // dive in
      }

      const sourceMapElement = element.getMetaProperty('sourceMap');
      const charStart = sourceMapElement.positionStart.get(2).toValue();
      const charEnd = sourceMapElement.positionEnd.get(2).toValue();
      const isWithinOffsetRange = this.offset >= charStart && (this.offset < charEnd || this.includeRightBound && this.offset <= charEnd);
      if (isWithinOffsetRange) {
        this.result.push(element);
        return undefined; // push to stack and dive in
      }

      return false; // skip entire sub-tree
    }
  }
});

// Finds the most inner node at the given offset.
// If includeRightBound is set, also finds nodes that end at the given offset.
// findAtOffset :: Number -> Element -> Element | Undefined
const findAtOffset = (0, _ramda.curry)((options, element) => {
  let offset;
  let includeRightBound;
  if ((0, _ramdaAdjunct.isNumber)(options)) {
    offset = options;
    includeRightBound = false;
  } else {
    offset = (0, _ramda.pathOr)(0, ['offset'], options);
    includeRightBound = (0, _ramda.pathOr)(false, ['includeRightBound'], options);
  }
  const visitor = Visitor({
    offset,
    includeRightBound
  });

  // @ts-ignore
  (0, _visitor.visit)(element, visitor);

  // @ts-ignore
  return (0, _ramda.last)(visitor.result);
});
var _default = findAtOffset;
exports.default = _default;
