// @swagger-api/apidom-core/cjs/traversal/parents.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _visitor = require("./visitor.cjs");
/* eslint-disable no-param-reassign */

const Visitor = (0, _stampit.default)({
  props: {
    parent: null,
    parentEdges: null
  },
  init() {
    this.parentEdges = new WeakMap();
  },
  methods: {
    ObjectElement(objectElement) {
      this.parentEdges.set(objectElement, this.parent);
      this.parent = objectElement;
    },
    ArrayElement(arrayElement) {
      this.parentEdges.set(arrayElement, this.parent);
      this.parent = arrayElement;
    },
    MemberElement(memberElement) {
      this.parentEdges.set(memberElement, this.parent);
      this.parent = memberElement;
    },
    enter(element) {
      this.parentEdges.set(element, this.parent);
    }
  }
});

/* eslint-enable */

// computes upwards edges from every child to its parent
// find :: Element -> WeakMap<Element, Element>
const parents = element => {
  const visitor = Visitor();
  (0, _visitor.visit)(element, visitor);
  return visitor.parentEdges;
};
var _default = parents;
exports.default = _default;
